### PR TITLE
removed an n * n * n problems with large one to one chains

### PR DESCRIPTION
**Pull request by Christian-B**

### CODE REVIEW

**Comment by Christian-B** [comment]
Ah found and fixed one edge case that was breaking test.

The one_to_one_groups previously included the same_chip_vertex_groups so the algorithm used the  ne_to_one_groups first and fell back on the same_chip_vertex_groups if required.

As one_to_one_groups now gives stops producing very large groups in some cases it returns incorrect too small one.

Fix is to concat the two group and use the concat first.

File: pacman/operations/placer_algorithms/one_to_one_placer.py [hunk 1]
```diff
@@ -86,31 +86,34 @@ class OneToOnePlacer(RadialPlacer):
...
         # Get which vertices must be placed on the same chip as another vertex
         same_chip_vertex_groups = get_same_chip_vertex_groups(machine_graph)
 
+        progress.update()
         # Work out the vertices that should be on the same chip by one-to-one
         # connectivity
         one_to_one_groups = group_vertices(
             machine_graph.vertices, functools.partial(
-                _find_one_to_one_vertices, graph=machine_graph))
+                _find_one_to_one_vertices, graph=machine_graph), 18)
```
Comment (alan-stokes) [inline pacman/operations/placer_algorithms/one_to_one_placer.py]: magic number

Comment (Christian-B) [reply pacman/operations/placer_algorithms/one_to_one_placer.py]: Agreed! 

I am going to try one alternative but I think that will be slower on machine with very large chains,

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py [hunk 2]
```diff
@@ -63,50 +64,67 @@ def group_vertices(vertices, same_group_as_function):
...
+    return same_chip_vertices
+
+
+def add_set(all_sets, new_set):
+    for a_set in all_sets:
+        intersection = new_set & a_set
+        if intersection:
+            all_sets.remove(a_set)
+            union = a_set | new_set
+            add_set(all_sets, union)
```
Comment (dkfellows) [inline pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: There should be some way to avoid writing this recursively.

Comment (dkfellows) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: Indeed, I think there ought to be a way to write this using a single pass.

Comment (Christian-B) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: Yep Done

Comment (dkfellows) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: Looks good

File: pacman/operations/placer_algorithms/one_to_one_placer.py [hunk 3]
```diff
@@ -88,31 +89,49 @@ class OneToOnePlacer(RadialPlacer):
 
     def __call__(self, machine_graph, machine):
 
+        # Iterate over vertices and generate placements
+        progress = ProgressBar(
+            machine_graph.n_vertices + 3, "Placing graph vertices")
```
Comment (alan-stokes) [inline pacman/operations/placer_algorithms/one_to_one_placer.py]: magic number 3. can we at least have a comment saying what these magic 3 steps are.

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py [hunk 4]
```diff
@@ -56,7 +57,7 @@ def get_same_chip_vertex_groups(graph):
         get_vertices_on_same_chip, graph=graph))
 
 
-def group_vertices(vertices, same_group_as_function):
+def group_vertices(vertices, same_group_as_function, cutoff=sys.maxsize):
```
Comment (alan-stokes) [inline pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: not in python doc

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py [hunk 5]
```diff
@@ -68,50 +69,70 @@ def group_vertices(vertices, same_group_as_function):
...
+    for group in groups:
+        for vertex in group:
+            same_chip_vertices[vertex] = group
+    for vertex in vertices:
+        if vertex not in same_chip_vertices:
+            same_chip_vertices[vertex] = {vertex}
+    return same_chip_vertices
+
+
+def add_set(all_sets, new_set):
```
Comment (alan-stokes) [inline pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: I might be being dense, but if the groups was a ordered set instead of a list, would this function become redundant, as you could just use update of the ordered set?

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py [hunk 6]
```diff
@@ -68,50 +69,70 @@ def group_vertices(vertices, same_group_as_function):
...
+            same_chip_as_vertices = OrderedSet(same_chip_as_vertices)
+            same_chip_as_vertices.add(vertex)
+            # Singletons on interesting and added later if needed
+            if len(same_chip_as_vertices) > 1:
+                add_set(groups, same_chip_as_vertices)
+    return groups
 
-    return same_chip_vertices
+
+def concat_all_groups(same_chip_as_vertices, same_chip_vertices, cutoff):
```
Comment (alan-stokes) [inline pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: this function isnt used anywhere. redundant code

File: pacman/operations/placer_algorithms/one_to_one_placer.py (hunk 3 above)
Comment (Christian-B) [reply pacman/operations/placer_algorithms/one_to_one_placer.py]: comment added

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py (hunk 4 above)
Comment (Christian-B) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: should I reply picky or good spot? either way done.

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py (hunk 5 above)
Comment (Christian-B) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: Added docs for the dense.

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py (hunk 4 above)
Comment (alan-stokes) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: well here's another one. now that the ide isnt complaining about lack of python docs, its complaining that its never used....... so redundant param

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py (hunk 5 above)
Comment (alan-stokes) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: fine, but could it be pushed into the orderedset fort encapsulation arguments? given its basically messing with sets.

File: pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py (hunk 4 above)
Comment (Christian-B) [reply pacman/utilities/algorithm_utilities/placer_algorithm_utilities.py]: yet left over from when I was using my own function vs set builtins

**Comment by Christian-B** [comment]
add_set can not be part of  orderedset  as its combines a new set into a list of existing sets.

See : PACMAN/unittests/utilities_tests/test_placer_algorithm_utilities.py for examples.

**Comment by alan-stokes** [comment]
fine, was just a thought.
